

#Aashish Tangnami


Objective:

To acheive the following text preprocessing steps with the given dataset - [emerson_self-reliance.txt]

1. Read data
2. Remove Punctuations
3. Tokenization
4. Remove Stopwords
5. Stemming
6. Lemmatize
7. Vecotrize



# Text Pre-Processing For ML Models

Data pre-processing is an essential step in building a Machine Learning model.
Natural Language Processing is branch of Data Science which deals with text data.
The following steps are need for transferring text from human language to machine readable format for further ML models.

The process of normalizating the text.
All the mentioned objectives has been explained in the below cells.



# Read Data

Very first step is to import raw data into the code base and there are multiple ways of doing so.

In [1]:
#importing libraries
import nltk
import pandas as pd
import re
import string

In [2]:
with open('emerson_self-reliance.txt') as f:
    tmp_text = f.read()

In [3]:
tmp_text

'Self-Reliance\nBy Ralph Waldo Emerson, 1841\n\n"Ne te quaesiveris extra."\n\n"Man is his own star; and the soul that can\nRender an honest and a perfect man,\nCommands all light, all influence, all fate;\nNothing to him falls early or too late.\nOur acts our angels are, or good or ill,\nOur fatal shadows that walk by us still."\n— Epilogue to Beaumont and Fletcher\'s Honest Man\'s Fortune.\n\nCast the bantling on the rocks,\nSuckle him with the she-wolf\'s teat,\nWintered with the hawk and fox,\nPower and speed be hands and feet.\n\nI read the other day some verses written by an eminent painter which were original and not conventional. Always the soul hears an admonition in such lines, let the subject be what it may. The sentiment they instill is of more value than any thought they may contain. To believe your own thought, to believe that what is true for you in your private heart is true for all men,-that is genius. Speak your latent conviction, and it shall be the universal sense; f

In [4]:

"""
Purpose of using inflect library is to convert the numbers in the text dataset into words.

We are trying to convert the numbers into words 

For example:  
text input : Abhiyan has 0 marks in dataware house assignment.
-> using inflect library function : number_to_words()
text ouptut: Abhiyan has no marks in dataware house assignment.

"""
# import the inflect library
import inflect

inflect_engine = inflect.engine()
 
# convert number into words
def convert_number_to_word(text):
    # split string into list of words
    raw_text = text.split()
    # initialise empty list
    new_text_data = []
 
    for word in raw_text:
        # if word is a digit, convert the digit
        # to numbers and append into the new_string list
        if word.isdigit():
            temp_word = inflect_engine.number_to_words(word)
            new_text_data.append(temp_word)
 
        # append the word as it is
        else:
            new_text_data.append(word)
 
    # join the words of new_string to form a string
    raw_text = ' '.join(new_text_data)
    return raw_text

In [5]:
new_text = convert_number_to_word(tmp_text)
new_text

'Self-Reliance By Ralph Waldo Emerson, one thousand, eight hundred and forty-one "Ne te quaesiveris extra." "Man is his own star; and the soul that can Render an honest and a perfect man, Commands all light, all influence, all fate; Nothing to him falls early or too late. Our acts our angels are, or good or ill, Our fatal shadows that walk by us still." — Epilogue to Beaumont and Fletcher\'s Honest Man\'s Fortune. Cast the bantling on the rocks, Suckle him with the she-wolf\'s teat, Wintered with the hawk and fox, Power and speed be hands and feet. I read the other day some verses written by an eminent painter which were original and not conventional. Always the soul hears an admonition in such lines, let the subject be what it may. The sentiment they instill is of more value than any thought they may contain. To believe your own thought, to believe that what is true for you in your private heart is true for all men,-that is genius. Speak your latent conviction, and it shall be the uni

### Observation:

1. We can notice that numeric value 1841 has been changed into word value.

# Lower Casing

This is foremost important step in the text cleaning so that all the text format for same text is treated as same. 

For eg. 'Aashish', 'AASHISH' and  'aashish' is treated as same. If not then the all three words are represented as three different words in the vector space model that results in more dimensions.

In [6]:
def lower_text(text):
  # custom function to lower the text format.
  # by default most of the tokenizer libraries lowers the case.
    text = text.lower()
    return text

lower_text = lower_text(new_text)
lower_text

'self-reliance by ralph waldo emerson, one thousand, eight hundred and forty-one "ne te quaesiveris extra." "man is his own star; and the soul that can render an honest and a perfect man, commands all light, all influence, all fate; nothing to him falls early or too late. our acts our angels are, or good or ill, our fatal shadows that walk by us still." — epilogue to beaumont and fletcher\'s honest man\'s fortune. cast the bantling on the rocks, suckle him with the she-wolf\'s teat, wintered with the hawk and fox, power and speed be hands and feet. i read the other day some verses written by an eminent painter which were original and not conventional. always the soul hears an admonition in such lines, let the subject be what it may. the sentiment they instill is of more value than any thought they may contain. to believe your own thought, to believe that what is true for you in your private heart is true for all men,-that is genius. speak your latent conviction, and it shall be the uni

### Observation:

1. All the text has been changed into lower cases. 

#Remove Punctuations

The process of removing punctuation from the text data is to standardize the text so that we can treat words such as 'Hi!' and 'Hi' as same word.

- "string.punctuation" in python contains the following punctuation symbols - "!"#$%&\'()*+,-./:;<=>?@[\\]^_{|}~`" 

In [7]:
def remove_punctuation(text):
    # function to remove the punctuation from the text.
    trans = str.maketrans('','', string.punctuation)
    return text.translate(trans)

punct_text = remove_punctuation(lower_text)
punct_text

'selfreliance by ralph waldo emerson one thousand eight hundred and fortyone ne te quaesiveris extra man is his own star and the soul that can render an honest and a perfect man commands all light all influence all fate nothing to him falls early or too late our acts our angels are or good or ill our fatal shadows that walk by us still — epilogue to beaumont and fletchers honest mans fortune cast the bantling on the rocks suckle him with the shewolfs teat wintered with the hawk and fox power and speed be hands and feet i read the other day some verses written by an eminent painter which were original and not conventional always the soul hears an admonition in such lines let the subject be what it may the sentiment they instill is of more value than any thought they may contain to believe your own thought to believe that what is true for you in your private heart is true for all menthat is genius speak your latent conviction and it shall be the universal sense for always the inmost beco

### Observation

1. All the special characters or symbols has been removed from the text.

2. Leading and ending spaces has been eliminated. 

3. On third line - has not been acknowledged as punctuation.

In [8]:
nltk.download('punkt')
from nltk.tokenize import sent_tokenize

corpus = nltk.sent_tokenize(lower_text)
corpus

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


['self-reliance by ralph waldo emerson, one thousand, eight hundred and forty-one "ne te quaesiveris extra."',
 '"man is his own star; and the soul that can render an honest and a perfect man, commands all light, all influence, all fate; nothing to him falls early or too late.',
 'our acts our angels are, or good or ill, our fatal shadows that walk by us still."',
 "— epilogue to beaumont and fletcher's honest man's fortune.",
 "cast the bantling on the rocks, suckle him with the she-wolf's teat, wintered with the hawk and fox, power and speed be hands and feet.",
 'i read the other day some verses written by an eminent painter which were original and not conventional.',
 'always the soul hears an admonition in such lines, let the subject be what it may.',
 'the sentiment they instill is of more value than any thought they may contain.',
 'to believe your own thought, to believe that what is true for you in your private heart is true for all men,-that is genius.',
 'speak your latent c

### Observation

1. We observed list of setences as corpus.

#Tokenization

The process of converting or splitting the sentences into words without any blanks or tabs is called tokenization.



In [9]:
# importing the tokenizer from nltk

from nltk.tokenize import word_tokenize
# nltk.download('punkt')

def text_tokenize(text):
    token_text = word_tokenize(text)
    return token_text


In [10]:
# tokenizing the punct text.
token_text = text_tokenize(punct_text)
token_text

['selfreliance',
 'by',
 'ralph',
 'waldo',
 'emerson',
 'one',
 'thousand',
 'eight',
 'hundred',
 'and',
 'fortyone',
 'ne',
 'te',
 'quaesiveris',
 'extra',
 'man',
 'is',
 'his',
 'own',
 'star',
 'and',
 'the',
 'soul',
 'that',
 'can',
 'render',
 'an',
 'honest',
 'and',
 'a',
 'perfect',
 'man',
 'commands',
 'all',
 'light',
 'all',
 'influence',
 'all',
 'fate',
 'nothing',
 'to',
 'him',
 'falls',
 'early',
 'or',
 'too',
 'late',
 'our',
 'acts',
 'our',
 'angels',
 'are',
 'or',
 'good',
 'or',
 'ill',
 'our',
 'fatal',
 'shadows',
 'that',
 'walk',
 'by',
 'us',
 'still',
 '—',
 'epilogue',
 'to',
 'beaumont',
 'and',
 'fletchers',
 'honest',
 'mans',
 'fortune',
 'cast',
 'the',
 'bantling',
 'on',
 'the',
 'rocks',
 'suckle',
 'him',
 'with',
 'the',
 'shewolfs',
 'teat',
 'wintered',
 'with',
 'the',
 'hawk',
 'and',
 'fox',
 'power',
 'and',
 'speed',
 'be',
 'hands',
 'and',
 'feet',
 'i',
 'read',
 'the',
 'other',
 'day',
 'some',
 'verses',
 'written',
 'by',
 'an

### Observation

1. Every sentence from the text dataset has been changed into tokens.

#Remove Stopwords

Stopwords are the most common words that do not carry important meaning thus are usually removed from texts.

In [11]:
# importing stopwords from nltk
from nltk.corpus import stopwords
nltk.download('stopwords')

def filtered_text(token_text):
    # function to filterout the stopwords from the tokens.
    stop_words = set(stopwords.words("english")) # set of stopwords.
    filtered_text = [token for token in token_text if token not in stop_words]
    return filtered_text

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [12]:
nonstop_text_sentence_token = filtered_text(corpus)
nonstop_text_sentence_token

['self-reliance by ralph waldo emerson, one thousand, eight hundred and forty-one "ne te quaesiveris extra."',
 '"man is his own star; and the soul that can render an honest and a perfect man, commands all light, all influence, all fate; nothing to him falls early or too late.',
 'our acts our angels are, or good or ill, our fatal shadows that walk by us still."',
 "— epilogue to beaumont and fletcher's honest man's fortune.",
 "cast the bantling on the rocks, suckle him with the she-wolf's teat, wintered with the hawk and fox, power and speed be hands and feet.",
 'i read the other day some verses written by an eminent painter which were original and not conventional.',
 'always the soul hears an admonition in such lines, let the subject be what it may.',
 'the sentiment they instill is of more value than any thought they may contain.',
 'to believe your own thought, to believe that what is true for you in your private heart is true for all men,-that is genius.',
 'speak your latent c

In [13]:
nonstop_text_token = filtered_text(token_text)
nonstop_text_token

['selfreliance',
 'ralph',
 'waldo',
 'emerson',
 'one',
 'thousand',
 'eight',
 'hundred',
 'fortyone',
 'ne',
 'te',
 'quaesiveris',
 'extra',
 'man',
 'star',
 'soul',
 'render',
 'honest',
 'perfect',
 'man',
 'commands',
 'light',
 'influence',
 'fate',
 'nothing',
 'falls',
 'early',
 'late',
 'acts',
 'angels',
 'good',
 'ill',
 'fatal',
 'shadows',
 'walk',
 'us',
 'still',
 '—',
 'epilogue',
 'beaumont',
 'fletchers',
 'honest',
 'mans',
 'fortune',
 'cast',
 'bantling',
 'rocks',
 'suckle',
 'shewolfs',
 'teat',
 'wintered',
 'hawk',
 'fox',
 'power',
 'speed',
 'hands',
 'feet',
 'read',
 'day',
 'verses',
 'written',
 'eminent',
 'painter',
 'original',
 'conventional',
 'always',
 'soul',
 'hears',
 'admonition',
 'lines',
 'let',
 'subject',
 'may',
 'sentiment',
 'instill',
 'value',
 'thought',
 'may',
 'contain',
 'believe',
 'thought',
 'believe',
 'true',
 'private',
 'heart',
 'true',
 'menthat',
 'genius',
 'speak',
 'latent',
 'conviction',
 'shall',
 'universal',

### Observations:

1. All the existing stopwords from the tokenized text data has been removed.

#Stemming

Stemming is a process of transforming or reducing a word to its root form.
for example the words "helping" and "helper" share root word "help".
Stemming enables you to focus on a word's core meaning rather than all the specifics of its usage. We are using stemmers available in NLTK.


Stemming has the drawback of altering the meaning of the root form of the word or preventing it from becoming a genuine english word.


In [14]:
from nltk.stem.porter import PorterStemmer
from nltk.tokenize import word_tokenize

stemmer = PorterStemmer()
 
# stem words in the list of tokenized words
def stem_words(text):
    stemmed_text = [stemmer.stem(word) for word in text]
    return stemmed_text


In [15]:
stemmed_text = stem_words(nonstop_text_token)
stemmed_text

['selfreli',
 'ralph',
 'waldo',
 'emerson',
 'one',
 'thousand',
 'eight',
 'hundr',
 'fortyon',
 'ne',
 'te',
 'quaesiveri',
 'extra',
 'man',
 'star',
 'soul',
 'render',
 'honest',
 'perfect',
 'man',
 'command',
 'light',
 'influenc',
 'fate',
 'noth',
 'fall',
 'earli',
 'late',
 'act',
 'angel',
 'good',
 'ill',
 'fatal',
 'shadow',
 'walk',
 'us',
 'still',
 '—',
 'epilogu',
 'beaumont',
 'fletcher',
 'honest',
 'man',
 'fortun',
 'cast',
 'bantl',
 'rock',
 'suckl',
 'shewolf',
 'teat',
 'winter',
 'hawk',
 'fox',
 'power',
 'speed',
 'hand',
 'feet',
 'read',
 'day',
 'vers',
 'written',
 'emin',
 'painter',
 'origin',
 'convent',
 'alway',
 'soul',
 'hear',
 'admonit',
 'line',
 'let',
 'subject',
 'may',
 'sentiment',
 'instil',
 'valu',
 'thought',
 'may',
 'contain',
 'believ',
 'thought',
 'believ',
 'true',
 'privat',
 'heart',
 'true',
 'menthat',
 'geniu',
 'speak',
 'latent',
 'convict',
 'shall',
 'univers',
 'sens',
 'alway',
 'inmost',
 'becom',
 'outmost—and',
 '

### Observation

1. Each word from the corpus are seperated as token.

#Lemmatize

Lemmatizing is similar to stemming in that it reduces words to their essence, but it gives you a full English word that stands alone as opposed to only a part of a word like "discoveri."

Unlike like stemming, lemmatizing reduces the plural word example: 'helps' and 'carves' becomes singular word as 'help' and 'scarf'



In [16]:
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize
nltk.download('wordnet')
nltk.download('omw-1.4')

lemmatizer = WordNetLemmatizer()

def lemmatize_word(text):
    # this function is to lemmatize tokenized words.
    lemmatize_text = [lemmatizer.lemmatize(word) for word in text]
    return lemmatize_text

lemmatized_text = lemmatize_word(nonstop_text_sentence_token)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [17]:
lemmatized_text

['self-reliance by ralph waldo emerson, one thousand, eight hundred and forty-one "ne te quaesiveris extra."',
 '"man is his own star; and the soul that can render an honest and a perfect man, commands all light, all influence, all fate; nothing to him falls early or too late.',
 'our acts our angels are, or good or ill, our fatal shadows that walk by us still."',
 "— epilogue to beaumont and fletcher's honest man's fortune.",
 "cast the bantling on the rocks, suckle him with the she-wolf's teat, wintered with the hawk and fox, power and speed be hands and feet.",
 'i read the other day some verses written by an eminent painter which were original and not conventional.',
 'always the soul hears an admonition in such lines, let the subject be what it may.',
 'the sentiment they instill is of more value than any thought they may contain.',
 'to believe your own thought, to believe that what is true for you in your private heart is true for all men,-that is genius.',
 'speak your latent c

### Observations

1. Lemmatization is preferred over stemming because lemmatization does a morphological analysis of the words.

2. Lemmatization considers the context and converts the word to its meaningful base form, whereas stemming just removes the last few characters, often leading to incorrect meanings and spelling errors.

#Vectorizing

Word vectorization is a methodology in NLP to map words or phrases from vocabulary to a corresponding vector of real numbers which used to find word predictions word similarities/semantics.

Basically process of converting words into numbers are called Vectorization.

We are using CountVectorizer
1. Simplest ways of doing text vectorization.

2. Creates a document term matrix of dummy variables if particular word appears in the document.

3. Count vectorizer will fit and learn the word vocabulary and try to create a document term matrix in which the individual cells denote the frequency of that word in a particular document, which is also known as term frequency and the columns are dedicated to each word in the corpus.

In [18]:
from sklearn.feature_extraction.text import CountVectorizer
# vectorizing the stemmed_text of the tokenized sentences.
vectorizer = CountVectorizer()
X = vectorizer.fit_transform(stemmed_text)
vectorizer.get_feature_names_out()

print(X.toarray())

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


In [19]:
# vectorizing the lemmatized text of the sentences as token from the dataset.
# this technique is used to find the relevance in meaning while doing text summarization in ML models.
vectorizer2 = CountVectorizer(analyzer='word', ngram_range=(2, 2))
X2 = vectorizer2.fit_transform(lemmatized_text)
vectorizer2.get_feature_names_out()
print(X2.toarray())

[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]


## SUMMARY

1. We studied and experienced how the text pre processing working and described its main steps including normalization, tokenization, stemming, lemmatization and vectorization.

2. We also studied difference between stemming and lemmatization how the context of words diminishes.

3. We have also studied the word tokens and sentence token to prepare text for further ML models in NLP.